File: src/ginger/hardware/led.py
"""8× WS2812B RGB LED strip via SPI (PCB V2.0, GRB order, SPI bus 0).

Uses MOSI (GPIO 10) at ~6.4 MHz to bit-bang WS2812B timing.
"""

import numpy as np
import spidev

_LED_COUNT = 8
_SPI_BUS = 0
_SPI_DEVICE = 0
_SPI_HZ = int(8 / 1.25e-6)  # 6.4 MHz


class LEDStrip:
    def __init__(self, count: int = _LED_COUNT, brightness: int = 255):
        self._count = count
        self._brightness = brightness
        # GRB order offsets
        self._g, self._r, self._b = 0, 1, 2
        self._buf = [[0, 0, 0]] * count  # stored as [G, R, B]

        self._spi = spidev.SpiDev()
        self._spi.open(_SPI_BUS, _SPI_DEVICE)
        self._spi.mode = 0
        self.clear()

    def set(self, index: int, r: int, g: int, b: int) -> None:
        scale = self._brightness / 255
        self._buf[index] = [
            round(g * scale),
            round(r * scale),
            round(b * scale),
        ]

    def set_all(self, r: int, g: int, b: int) -> None:
        for i in range(self._count):
            self.set(i, r, g, b)

    def show(self) -> None:
        flat = np.array([v for pixel in self._buf for v in pixel], dtype=np.uint8)
        tx = np.zeros(len(flat) * 8, dtype=np.uint8)
        for bit in range(8):
            tx[7 - bit::8] = ((flat >> bit) & 1) * 0x78 + 0x80
        self._spi.xfer(tx.tolist(), _SPI_HZ)

    def clear(self) -> None:
        self.set_all(0, 0, 0)
        self.show()

    def close(self) -> None:
        self.clear()
        self._spi.close()

File: src/hal/led.rs
//! 8× WS2812B RGB LED strip via SPI (SPI bus 0, GRB order).
//!
//! Uses MOSI at ~6.4 MHz to encode WS2812B bit timing:
//!   bit=1 → 0xF8, bit=0 → 0x80  (one SPI byte per WS2812B bit)

use rppal::spi::{Bus, Mode, SlaveSelect, Spi};

use crate::Result;

pub const LED_COUNT: usize = 8;
const SPI_HZ: u32 = 6_400_000; // 8 bits / 1.25 µs ≈ 6.4 MHz

pub struct LedStrip {
    spi: Spi,
    buf: [[u8; 3]; LED_COUNT], // stored as [G, R, B]
    brightness: u8,
}

impl LedStrip {
    pub fn new(brightness: u8) -> Result<Self> {
        let spi = Spi::new(Bus::Spi0, SlaveSelect::Ss0, SPI_HZ, Mode::Mode0)?;
        let mut strip = Self {
            spi,
            buf: [[0; 3]; LED_COUNT],
            brightness,
        };
        strip.show()?;
        Ok(strip)
    }

    pub fn set(&mut self, index: usize, r: u8, g: u8, b: u8) {
        let scale = self.brightness as f32 / 255.0;
        self.buf[index] = [
            (g as f32 * scale).round() as u8,
            (r as f32 * scale).round() as u8,
            (b as f32 * scale).round() as u8,
        ];
    }

    pub fn set_all(&mut self, r: u8, g: u8, b: u8) {
        for i in 0..LED_COUNT {
            self.set(i, r, g, b);
        }
    }

    pub fn show(&mut self) -> Result<()> {
        let flat: Vec<u8> = self.buf.iter().flat_map(|p| p.iter().copied()).collect();
        let mut tx = vec![0u8; flat.len() * 8];
        for (byte_idx, &byte) in flat.iter().enumerate() {
            for bit in 0..8u8 {
                let spi_byte = if (byte >> (7 - bit)) & 1 == 1 {
                    0xF8
                } else {
                    0x80
                };
                tx[byte_idx * 8 + bit as usize] = spi_byte;
            }
        }
        self.spi.write(&tx)?;
        Ok(())
    }

    pub fn clear(&mut self) -> Result<()> {
        self.set_all(0, 0, 0);
        self.show()
    }
}
